### PR TITLE
Fix soundness

### PR DESCRIPTION
**Pull request by allanrenucci**

### CODE REVIEW

**Comment by sjrd** [comment]
I'd like to point out that those `protected[this]` members *are* (likely) unsound. AFAICT, they're very similar to `thisCollection` and `toCollection` in the old collections, which *do* cause `ClassCastException`s if they're not systematically overridden by subclasses.

For example, see the old issue https://github.com/scala-js/scala-js/issues/843 and its fix https://github.com/scala-js/scala-js/pull/851/files.

**Comment by odersky** [comment]
@sjrd Yes, that's my sentiment as well. It needs to be documented thoroughly.

**Comment by julienrf** [comment]
I’ve just pushed a commit to get `withFilter` working. It is implemented on top of the unsound `IterableCC`-like types, but I think if we want to get it right we’d really have to have one `WithFilter` implementation per subclass of `Iterable`.

**Comment by julienrf** [comment]
I’ve also fixed `ArrayDeque` so that `Stack` and `Queue` can benefit from its overrides.

**Comment by julienrf** [comment]
Some `junit/test`s are still failing with Dotty.

File: .travis.yml [hunk 1]
```diff
@@ -1,7 +1,8 @@
 language: scala
 jdk: oraclejdk8
 script:
-  - sbt ++2.12.4 memoryBenchmark/compile test ++2.13.0-M2 timeBenchmark/compile test ++0.7.0-RC1 test
+  - sbt ++2.12.4 memoryBenchmark/compile test ++2.13.0-M2 timeBenchmark/compile test
+  - sbt ++0.8.0-bin-20180323-5be1360-NIGHTLY test junit/test:compile
```
[the remark this answers is not in the document]
Comment (allanrenucci) [reply .travis.yml]: @julienrf I you managed to fix the `collections-contrib` with Dotty, you can add them to the CI tests

[the remark this answers is not in the document]
Comment (julienrf) [reply .travis.yml]: There is still a compiler crash when compiling `MapDecoratorTest.scala`, but at least `collections-contribJVM/compile` works with Dotty.

[the remark this answers is not in the document]
Comment (allanrenucci) [reply .travis.yml]: Ok thanks! I'll look into the crash

**Comment by julienrf** [comment]
This PR has been ported to scala/scala but I leave it open here because we will eventually have to also port the part that touches the `collections-contrib` project.

**Comment by julienrf** [comment]
Superseded by #561 and scala/scala#6508

**Pull request closed without merging**